File: golang/107.binary-tree-level-order-traversal-ii.go
/*
 * @lc app=leetcode id=107 lang=golang
 *
 * [107] Binary Tree Level Order Traversal II
 */

package golang

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
	res := make([][]int, 0)
	if root == nil {
		return res
	}
	helpLevel(root, 0, &res)
	for i := 0; i < len(res)/2; i++ {
		res[i], res[len(res)-i-1] = res[len(res)-i-1], res[i]
	}
	return res
}

func helpLevel(root *TreeNode, depth int, res *[][]int) {
	if depth == len(*res) {
		*res = append(*res, []int{})
	}
	(*res)[depth] = append((*res)[depth], root.Val)
	if root.Left != nil {
		helpLevel(root.Left, depth+1, res)
	}
	if root.Right != nil {
		helpLevel(root.Right, depth+1, res)
	}
}

// @lc code=end

File: typescript/src/107.binary-tree-level-order-traversal-ii.ts
/*
 * @lc app=leetcode id=107 lang=typescript
 *
 * [107] Binary Tree Level Order Traversal II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrderBottom(root: TreeNode | null): number[][] {
    let trees: (TreeNode | null)[] = [root]
    return helpLevelOrderBottomRecursive(trees).slice(1)
};

function helpLevelOrderBottomRecursive(trees: (TreeNode | null)[]): number[][] {
    if (trees.length == 0) {
        return []
    }
    let nextTrees: (TreeNode | null)[] = []
    let nums: number[] = []
    trees.forEach(node => {
        if (node !== null) {
            nums.push(node.val)
            nextTrees.push(node.left, node.right)
        }
    })
    let result = helpLevelOrderBottomRecursive(nextTrees)
    result.push(nums)
    return result
}
// @lc code=end

